test/scripts/hack_task_test.rb

```diff
@@ -0,0 +1,68 @@
+require "test_helper"
+
+class HackTaskTest < ActiveSupport::TestCase
+  # TODO: Try to get these values dynamically.
+  FRAMEWORK_PACKAGES = [
+    "bullet_train",
+    "bullet_train-api",
+    "bullet_train-fields",
+    "bullet_train-has_uuid",
+    "bullet_train-incoming_webhooks",
+    "bullet_train-integrations",
+    "bullet_train-integrations-stripe",
+    "bullet_train-obfuscates_id",
+    "bullet_train-outgoing_webhooks",
+
+    # TODO: We need to add this with bin/hack --link
+    # "bullet_train-roles",
+
+    "bullet_train-scope_questions",
+    "bullet_train-scope_validator",
+    "bullet_train-sortable",
+    "bullet_train-super_load_and_authorize_resource",
+    "bullet_train-super_scaffolding",
+    "bullet_train-themes",
+    "bullet_train-themes-light",
+    "bullet_train-themes-tailwind_css",
+  ]
+
+  def setup
+    @original_gemfile_lines = File.open("Gemfile").readlines
+    @temp_gemfile_lines = File.open("test/support/temp_gemfile.rb").readlines
+
+    @current_bullet_train_gem_version = @original_gemfile_lines.find do |line|
+      line.match?('gem "bullet_train"')
+    end
+
+    # Because these tests are being run against a task in the `bullet_train` gem,
+    # we ensure that the original `bullet_train` gem the developer is using stays in the temp Gemfile.
+    # We can afford to skip checking `path: ...` for this gem in the individual test cases
+    # because the `--link` logic is mostly covered for the other Bullet Train gems that are in the Gemfile,
+    # including ones that aren't originally there but are added with `bin/hack --link`.
+    @temp_gemfile_lines = @temp_gemfile_lines.map do |line|
+      line.match?('gem "bullet_train"') ? @current_bullet_train_gem_version : line
+    end
+
+    # Replace the Gemfile with the temp file.
+    File.write("Gemfile", @temp_gemfile_lines.join)
```

Comment:
I don't quite understand what this `setup` routine is supposed to do. If I insert a `binding.pry` at the end, after we're re-written `Gemfile` I'm not seeing that any actual change has been made. The line that says `gem "bullet_train"` prior to `setup` still says `gem "bullet_train"` after `setup` has run.

So I think that the main thing `setup` is doing currently is capturing the pre-test state of `Gemfile` so that we can restore it. If that's the case (and I'm not missing something) it might be simpler and easier to follow if we just explicitly made a copy of `Gemfile` and then restored it.

Running this test is also leaving my local repo in a dirty state, with uncommitted changes to `Gemfile.lock`, so we may need to restore that one as well.

Something like:

```ruby
def setup
  `cp Gemfile Gemfile.pre_test`
  `cp Gemfile.lock Gemfile.lock.pre_test`
end

def teardown
  `mv Gemfile.pre_test Gemfile`
  `mv Gemfile.lock.pre_test Gemfile.lock`
end
```

Reply:
Oh, I think maybe I understand why `setup` is the way it is.

In CI, due to the way `bin/checkout-and-link-core-repo` works the pre-test state of `Gemfile` would be saying:

```ruby
gem "bullet_train", path: "core/bullet_train"
```

Instead of just:

```ruby
gem "bullet_train"
```

And we'd want to keep pointing to `core/bullet_train` instead of `local/bullet_train-core/bullet_train` to ensure that we're testing against whichever branch of `core` we may have determined to use via our branch-matching process. Without that `setup` process we'd end up testing against the `main` branch, which is what we'd get by default when running `bin/hack --link`.

**BUT** since we don't _already_ have anything in `local/bullet_train-core` we end up skipping this test entirely, which is probably not what we want in CI.

I think we need to remove that conditional and try to figure out a different way around whatever is causing things to fail.

Reply:
Just had a thought about this. I kinda think that we can safely assume that `bin/hack --link` should always be working against a "clean" `Gemfile`. That is, one that starts out pointing to released gems, and not to random local paths.

If we're comfortable with that assumption then can we just copy `test/support/temp_gemfile.rb` over the top of `Gemfile` in `setup` and then restore the original state in `teardown`?

```ruby
def setup
  `cp Gemfile Gemfile.pre_test`
  `cp Gemfile.lock Gemfile.lock.pre_test`
  `cp test/support/temp_gemfile.rb Gemfile`
  `bundle install`
end

def teardown
  `mv Gemfile.pre_test Gemfile`
  `mv Gemfile.lock.pre_test Gemfile.lock`
  `bundle install`
end
```